pkg/secrets/s3.go

```diff
@@ -0,0 +1,122 @@
+package secrets
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	yamlParse "gopkg.in/yaml.v2"
+	"reflect"
+	"strings"
+)
+
+type S3Secret struct {
+	region   string
+	bucket   string
+	filepath string
+	key      string
+}
+
+type S3Decrypter struct {
+	secretConfig string
+}
+
+func NewS3Decrypter(secretConfig string) *S3Decrypter {
+	return &S3Decrypter{secretConfig}
+}
+
+func (s3 *S3Decrypter) Decrypt() ([]byte, error) {
+	s3Secret, err := parseS3SecretConfig(s3.secretConfig)
+	if err != nil {
+		return []byte{}, err
+	}
+	secret, err := s3Secret.fetchSecret()
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return secret, nil
+}
+
+func parseS3SecretConfig(secretConfig string) (S3Secret, error) {
+	if !strings.HasPrefix(secretConfig, "encrypted:") {
+		return S3Secret{}, fmt.Errorf("decrypt called on secret without encrypted syntax: %s", secretConfig)
+	}
+
+	var s3Secret S3Secret
+	segments := strings.Split(secretConfig, "!")
+	if len(segments) < 2 {
+		return S3Secret{}, fmt.Errorf("bad format for secret syntax: %s", secretConfig)
+	}
+	for _, element := range segments {
+		kv := strings.Split(element, ":")
+		if len(kv) < 2 {
+			return S3Secret{}, fmt.Errorf("bad format for key-value pair in secret syntax: %s", element)
+		}
+		switch kv[0] {
+		case "encrypted":
+			// do nothing
+		case "r":
+			s3Secret.region = kv[1]
+		case "b":
+			s3Secret.bucket = kv[1]
+		case "f":
+			s3Secret.filepath = kv[1]
+		case "k":
+			s3Secret.key = kv[1]
+		default:
+			return S3Secret{}, fmt.Errorf("invalid key in secret syntax: %s", kv[0])
+		}
+	}
+	return s3Secret, nil
+}
+
+func (secret *S3Secret) fetchSecret() ([]byte, error) {
+	sess, err := session.NewSession(&aws.Config{
+		Region: &secret.region,
+	})
+
+	downloader := s3manager.NewDownloader(sess)
+
+	contents := aws.NewWriteAtBuffer([]byte{})
```

Comment:
Allocate a new bytes.Buffer unless this function specifically requires a byte array

Reply:
That function takes a byte array and I couldn't find a way to use byte.Buffer, but I don't know what I'm doing, so...

Reply:
I looked at the func, it only accepts a byte array.